internal/suppliers/car_ciditerator.go

```diff
@@ -0,0 +1,57 @@
+package suppliers
+
+import (
+	"context"
+	"io"
+
+	"github.com/ipfs/go-cid"
+	"github.com/ipld/go-car/v2"
+	"github.com/ipld/go-car/v2/blockstore"
+)
+
+type carCidIterator struct {
+	closer   io.Closer
+	cancel   context.CancelFunc
+	cidsChan <-chan cid.Cid
+	errChan  <-chan error
+}
+
+func newCarCidIterator(path string, opts ...car.ReadOption) (*carCidIterator, error) {
+	robs, err := blockstore.OpenReadOnly(path, opts...)
+	if err != nil {
+		return nil, err
+	}
+	errChan := make(chan error)
+	ctx, cancel := context.WithCancel(context.Background())
+	ctx = blockstore.WithAsyncErrorHandler(ctx, func(err error) { errChan <- err })
+	cidsChan, err := robs.AllKeysChan(ctx)
+	if err != nil {
+		cancel()
+		_ = robs.Close()
+		return nil, err
+	}
+
+	return &carCidIterator{
+		closer:   robs,
+		cancel:   cancel,
+		cidsChan: cidsChan,
+		errChan:  errChan,
+	}, nil
+}
+
+func (c *carCidIterator) Next() (cid.Cid, error) {
```

Comment:
@willscott If multihash iterator is not a blocker for this PR, I'd like to work on that in the separate PR please. For now I have left TODOs in relevant places.

Reply:
Continuing with this until CARv2 index work is prepared to allow reconstruction of multihashes from CARv2 index alone rather than traversing the blocks.